src/provider/model/Model.ts

```diff
@@ -128,21 +131,32 @@ export class Model {
     }
 
     private async onWindowCreated(identity: Identity, manifestUrl: string): Promise<void> {
+        // Registration is asynchronous and sensitive to race conditions. We use a deferred promise
+        // to signal to other sensitive operations that it is safe to proceed.
+        const [pendingRegistration, clearPending] = deferredPromise();
+        this._pendingRegistrations.set(getId(identity), pendingRegistration);
+
         const apps = await this._directory.getAllApps();
         const appInfoFromDirectory = apps.find(app => app.manifest.startsWith(manifestUrl));
 
+        const id: string = getId(identity);
         if (!appInfoFromDirectory) {
             // If the app is not in directory we ignore it. We'll add it to the model if and when it connects to FDC3
+            clearPending();
+            this._pendingRegistrations.delete(getId(identity));
             return;
-        }
-
-        const id: string = getId(identity);
-        if (this._windowsById[id]) {
+        } else if (this._windowsById[id]) {
             console.info(`Ignoring window created event for ${id} - window was already registered`);
+            clearPending();
+            this._pendingRegistrations.delete(getId(identity));
             return;
+        } else {
+            this.registerWindow(appInfoFromDirectory, identity, true);
         }
 
-        this.registerWindow(appInfoFromDirectory, identity, true);
+        // Registration finished, allow any other sensitive operations to proceed
+        clearPending();
```

Comment:
Suggest resolve rather than clear, to stick with vocab we already have for promises

Reply:
I flipped back and forth a couple of times on that. I'll put it back to resolve